Add Config Flow to LG Netcast

splinter98 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
[Turn on action][turn_on_action_doc] for turning the TV on via a custom script within `configuration.yaml` has been changed to use a custom automation trigger `lg_netcast.turn_on` or a device trigger that can be setup from the UI.

[turn_on_action_doc]: https://www.home-assistant.io/integrations/lg_netcast/#advanced-configuration

Example trigger automation:

```yaml
automaiton:
  trigger:
    - platform: lg_netcast.turn_on
      entity_id: media_player.lg_netcast_smart_tv
  action:
    - service: wake_on_lan.send_magic_packet
      data:
        mac: AA-BB-CC-DD-EE-FF
```

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add config flow to LG Netcast 
Also adds the ability to discover existing devices on the network.

This does not enable auto discovery as in my testing, every ping out to the network causes the device PIN to display on the TV, which becomes irritating.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30110#issue-2022193600

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Drafteed commented:
Hello! Great first contribution 👍

Can you split this PR into two (first config flow and then autodiscovery) to make the code review easier, as the [documentation recommends](https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr)?

splinter98 commented:
> The existing YAML configuration must be imported and deprecated (with generating a system issue).

I can add a YAML migration flow, the only parameter that as far as I can tell is unsupported in the config flow will be the `turn_on_action` which accepts a custom script so I'm assuming we will need to then deprecate this parameter?

Also, the current implementation only supports a single manual tv instance through the flow as the currently used API/library doesn't expose a unique id.

Thinking about it, I could perhaps use a try and use a similar endpoint that the auto discovery would share as that does expose the same UUID that the discovery flow would use, and fail if it's not found (I don't know if it would always be exposed or not) - but that logic may be better suited in the pylgnetcast library?

Drafteed commented:
Sorry for the late answer - lack of time. 

> I can add a YAML migration flow, the only parameter that as far as I can tell is unsupported in the config flow will be the turn_on_action which accepts a custom script so I'm assuming we will need to then deprecate this parameter?

I think we can do the same as it was done in a similar integration https://github.com/home-assistant/core/pull/64117.

> but that logic may be better suited in the pylgnetcast library?

I would like to rewrite it using a more modern API, but everything is complicated by the fact that the only LG Netcast TV I have is at my parents' house, and I don't live there. This also makes review PR difficult for me. If you have enthusiasm and time, then [PR is welcome](https://github.com/Drafteed/python-lgnetcast). I can also make you a maintainer in the upstream lib. 

In addition, such a big change, I suggest you add yourself as `codeowner` there.

splinter98 commented:
> Sorry for the late answer - lack of time.

No Problem at all

> I think we can do the same as it was done in a similar integration #64117.

Yes - I've been reading how other integrations handle this and I had stumbled across this and came to a similar conclusion.

> I would like to rewrite it using a more modern API, but everything is complicated by the fact that the only LG Netcast TV I have is at my parents' house, and I don't live there. This also makes review PR difficult for me. If you have enthusiasm and time, then [PR is welcome](https://github.com/Drafteed/python-lgnetcast). I can also make you a maintainer in the upstream lib.

Yes, I saw that you were interested in rewriting the library, I'm more than happy to take a first pass, and as I own a 2012 LG Netcast TV it certainly simplifies development on my side.

> In addition, such a big change, I suggest you add yourself as `codeowner` there.

Absolutely, I will do that when I've finished handling the YAML import step.

splinter98 commented:
I've now added the YMAL migration flow and implemented the `turn_on` as a device trigger. As that is a manual conversion, I wasn't sure if I should add a dedicated issue specifically if the YAML migration detects the old `turn_on_action` is detected in the YAML or is the current issue sufficient?

Drafteed commented:
> As that is a manual conversion, I wasn't sure if I should add a dedicated issue specifically if the YAML migration detects the old turn_on_action is detected in the YAML or is the current issue sufficient?

Honestly, I don't know what's best practice to handle this rare case. Personally, I think that current (deprecated yaml) issue is sufficient.

Drafteed commented on homeassistant/components/lg_netcast/helpers.py:
```diff
@@ -0,0 +1,110 @@
+class NetcastDetails(TypedDict):
+    """Netcast TV Details."""
+
+    uuid: str
+    model_name: str
+    friendly_name: str
+
+
+async def async_discover_netcast_details(
+    hass: HomeAssistant, client: LgNetCastClient
+) -> NetcastDetails:
+    """Discover UUID and Model Name from Netcast Tv."""
+    # We're using UDAP to retrieve this information, which requires a specific User-Agent
+    client.HEADER = {**LgNetCastClient.HEADER, "User-Agent": "UDAP/2.0"}
+
+    try:
+        resp = await hass.async_add_executor_job(
+            partial(client._send_to_tv, payload={"target": "rootservice.xml"}),  # pylint: disable=protected-access
+            "data",
+        )
+    except RequestException as err:
+        raise LGNetCastDetailDiscoveryError(
+            f"Error in connecting to {client.url}"
+        ) from err
+
+    if resp.status_code != 200:
+        raise LGNetCastDetailDiscoveryError(
+            "Invalid response ({resp.status_code}) from: {resp.url}"
+        )
+
+    try:
+        tree = DET.fromstring(resp.text.encode("utf-8"))
+    except DET.ParseError as err:
+        raise LGNetCastDetailDiscoveryError("Invalid XML") from err
+
+    return {
+        "uuid": tree.findtext("device/uuid"),
+        "model_name": tree.findtext("device/modelName"),
+        "friendly_name": tree.findtext("device/friendlyName"),
+    }
```
Consider moving this to the backend. I invited you as a collaborator.

splinter98 replied:
I've moved the code to the backend and pushed a new release and updated the dependencies

Drafteed commented on homeassistant/components/lg_netcast/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+
+            errors[CONF_HOST] = "invalid_host"
+
+        user_input = user_input or {}
```
Why is this statement needed?

splinter98 replied:
Good spot, it's left over from the auto discovery logic that I removed from the initial commit, where `CONF_HOST` becomes an optional field and I was setting the default based on the value from `user_input` if available (otherwise it's empty). But it is unnecessary for the basic config flow implementation. I will remove it

splinter98 replied:
This line is now removed

Drafteed reviewed (commented):
I was able to quickly test this - everything works great!
I left a couple of comments - check them out.
Overall it looks good to me - great job 👍

Drafteed commented:
The code looks great to me. I'll test it out on my LG Netcast TV on this weekend.

Drafteed commented on homeassistant/components/lg_netcast/media_player.py:
```diff
@@ -79,18 +109,36 @@ class LgTVDevice(MediaPlayerEntity):
         self._source_names = []
+        if unique_id is not None:
+            self._attr_unique_id = unique_id
+            self._attr_device_info = DeviceInfo(
+                identifiers={(DOMAIN, unique_id)},
+                manufacturer=ATTR_MANUFACTURER,
+                name=f"{ATTR_MANUFACTURER} Netcast",
+            )
```
We should provide `MODEL_NAME` in the `model` field of device info. In the `name` field we can also provide `MANUFACTURER_NAME + MODEL_NAME` or we can use device `friendly_name`. In case of export, we need to use the name from YAML to avoid breaking changes.

See example: https://github.com/home-assistant/core/blob/e0490a3ade04fdc3fcbcacb16ce56ad308408f31/homeassistant/components/braviatv/entity.py#L24-L28

or 
https://github.com/home-assistant/core/blob/e0490a3ade04fdc3fcbcacb16ce56ad308408f31/homeassistant/components/cast/media_player.py#L291-L295

Drafteed replied:
We also should set `_attr_has_entity_name = True` and `_attr_name = None` in the `MediaPlayerEntity`. See https://developers.home-assistant.io/blog/2022/07/10/entity_naming/

splinter98 replied:
> We should provide `MODEL_NAME` in the `model` field of device info. In the `name` field we can also provide `MANUFACTURER_NAME + MODEL_NAME` or we can use device `friendly_name`. In case of export, we need to use the name from YAML to avoid breaking changes.

I was already generating a `CONF_NAME` but wasn't exposing that through - so I've now switched that to using just the friendly name and there was already a check to check to use the YAML name, which should now be picked up.



> We also should set `_attr_has_entity_name = True` and `_attr_name = None` in the `MediaPlayerEntity`. See https://developers.home-assistant.io/blog/2022/07/10/entity_naming/

Done

Drafteed commented on homeassistant/components/lg_netcast/strings.json:
```diff
@@ -0,0 +1,33 @@
+    "step": {
+      "user": {
+        "description": "Ensure that your TV is turned on before trying to set it up.\nIf you leave the host empty, discovery will be used to find devices.",
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]"
+        }
```
Let`s add host filed description.

Example:
https://github.com/home-assistant/core/blob/e0490a3ade04fdc3fcbcacb16ce56ad308408f31/homeassistant/components/braviatv/strings.json#L6-L11

splinter98 replied:
This has been added

Drafteed commented on homeassistant/components/lg_netcast/config_flow.py:
```diff
@@ -0,0 +1,160 @@
+            try:
+                details = await async_discover_netcast_details(self.hass, self.client)
+            except LGNetCastDetailDiscoveryError:
+                return self.async_abort(reason="cannot_connect")
+            unique_id = details["uuid"]
+            if unique_id is None:
+                return self.async_abort(reason="invalid_host")
+            self.device_config[CONF_ID] = unique_id
+            if CONF_NAME not in self.device_config:
+                model_name = details["model_name"]
+                friendly_name = details["friendly_name"] or DEFAULT_NAME
+                self.device_config[CONF_NAME] = (
+                    f"{friendly_name} ({model_name})" if model_name else friendly_name
+                )
+
+        await self.async_set_unique_id(self.device_config[CONF_ID])
+        self._abort_if_unique_id_configured(
+            updates={CONF_HOST: self.device_config[CONF_HOST]}
+        )
+
+        try:
+            await self.hass.async_add_executor_job(
+                self.client._get_session_id  # pylint: disable=protected-access
+            )
+            return await self.async_create_device()
+        except AccessTokenError:
+            if user_input is not None:
+                errors[CONF_ACCESS_TOKEN] = "invalid_access_token"
+        except SessionIdError:
+            errors["base"] = "cannot_connect"
+
+        return self.async_show_form(
+            step_id="authorize",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(CONF_ACCESS_TOKEN): vol.All(str, vol.Length(max=6)),
+                }
+            ),
+            errors=errors,
+        )
```
When I re-added the device, after entering the IP address, the integration asks to enter an access token, but the TV does not show it. To see it I had to enter `0000`. Maybe we should do this automatically like the Bravia integration does (`self.client.pair` trying to connect with `0000` PIN which causes a popup on the TV)?

https://github.com/home-assistant/core/blob/e0490a3ade04fdc3fcbcacb16ce56ad308408f31/homeassistant/components/braviatv/config_flow.py#L163-L166

splinter98 replied:
When no access token is provided it is sending an auth token request already, however I have seen similar issues to what you've described. I think it's related to some timers in the TV as it is seemingly random. The way I had intended to handle this was by submitting an empty access token to force the auth to happen again, which in my testing usually works. But I note that this wasn't obvious this was possible.

I've now implemented the ability for the flow to regularly send the auth request every second whilst we're waiting for the access token to improve the user experience, which should handle both the prompt not appearing (or briefly appearing and then immediately disappearing), but also removes the time sensitivity that was there previously.

If you find that you still need to send an invalid token I can modify this to do so.

Drafteed reviewed (commented):
Everything works fine, I left a couple of comments.

splinter98 committed:
Add Config Flow to lg_netcast

splinter98 committed:
Add YAML import to Lg Netcast ConfigFlow

Deprecates YAML config support

splinter98 committed:
Add LG Netcast Device triggers for turn_on action

splinter98 committed:
Add myself to LG Netcast codeowners

splinter98 committed:
Remove unnecessary user_input validation check.

splinter98 committed:
Move netcast discovery logic to the backend

splinter98 committed:
Use FlowResultType Enum for tests

splinter98 committed:
Mock pylgnetcast.query_device_info instead of _send_to_tv

splinter98 committed:
Refactor lg_netcast client discovery, simplify YAML import

splinter98 committed:
Simplify CONF_NAME to use friendly name

Fix: Use Friendly name for Name

splinter98 committed:
Expose model to DeviceInfo

splinter98 committed:
Add test for testing YAML import when not TV not online

splinter98 committed:
Switch to entity_name for LGTVDevice

splinter98 committed:
Add data_description to host field in user step

splinter98 committed:
Wrap try only around _get_session_id

splinter98 committed:
Send regular request for access_token to ensure it display on the TV

splinter98 commented:
I've now implemented the fixes - I will apologise for all the commits being repushed, I had realised that they had the wrong email attached them and decided they should be fixed before this got merged in. They are exactly the same as they were previously. The new commits start from [c1c580d](https://github.com/home-assistant/core/pull/104913/commits/c1c580d4ce303de18c5d27d8422414bb2d990899) onwards

splinter98 committed:
Stop displaying access token when flow is aborted

Drafteed reviewed (approved):
Looks good to me 👍
Tested, everything works.

emontnemery commented on homeassistant/components/lg_netcast/config_flow.py:
```diff
@@ -0,0 +1,213 @@
+
+        try:
+            self._async_abort_entries_match({CONF_ID: self.device_config[CONF_ID]})
+        except AbortFlow as err:
+            async_create_issue(
+                self.hass,
+                DOMAIN,
+                "deprecated_yaml_import_issue_already_configured",
+                breaks_in_ha_version="2024.6.0",
+                is_fixable=False,
+                issue_domain=DOMAIN,
+                severity=IssueSeverity.WARNING,
+                translation_key="deprecated_yaml_import_issue_already_configured",
+                translation_placeholders={
+                    "domain": DOMAIN,
+                    "interation_title": "LG Netcast",
+                },
+            )
+            raise err
+
+        async_create_issue(
+            self.hass,
+            HOMEASSISTANT_DOMAIN,
+            f"deprecated_yaml_{DOMAIN}",
+            breaks_in_ha_version="2024.6.0",
+            is_fixable=False,
+            issue_domain=DOMAIN,
+            severity=IssueSeverity.WARNING,
+            translation_key="deprecated_yaml",
+            translation_placeholders={
+                "domain": DOMAIN,
+                "integration_title": "LG Netcast",
+            },
+        )
```
The way in which we distinguish between these two cases is not correct, because it makes no difference to the user.
Instead, we should distinguish between:
1. The import is done, and the user should remove YAML
2. Something went wrong when doing import (if this can happen, for example in the authorize step, and we can't leave the flow import for the user to finish)

splinter98 replied:
I've refactored this and now raising issues if we get an error - and only raising the deprecated_yaml issue if we succeed (or fail because it's already migrated)

emontnemery commented on homeassistant/components/lg_netcast/config_flow.py:
```diff
@@ -0,0 +1,213 @@
+            errors=errors,
+        )
+
+    @callback
```
```suggestion
```

emontnemery commented on homeassistant/components/lg_netcast/const.py:
```diff
@@ -1,3 +1,15 @@
+from datetime import timedelta
+from typing import Final
+
+ATTR_MANUFACTURER: Final = "LG"
+ATTR_MODEL_NAME = "model_name"
+ATTR_FRIENDLY_NAME = "friendly_name"
+ATTR_UUID = "uuid"
```
Please remove these constants, they're only used once, in config_flow.py

splinter98 replied:
Done

emontnemery commented on homeassistant/components/lg_netcast/const.py:
```diff
@@ -1,3 +1,15 @@
+
 DOMAIN = "lg_netcast"
+
+DISPLAY_ACCESS_TOKEN_INTERVAL = timedelta(seconds=1)
```
Move this to config_flow.py

emontnemery commented on homeassistant/components/lg_netcast/helpers.py:
```diff
@@ -0,0 +1,94 @@
+    if (device := device_reg.async_get(device_id)) is None:
+        raise ValueError(f"Device {device_id} is not a valid {DOMAIN} device.")
+
+    return device
+
+
+@callback
+def async_get_client_by_device_entry(
+    hass: HomeAssistant, device: DeviceEntry
+) -> LgNetCastClient:
+    """Get LG Netcast from Device Registry by device entry.
+
+    Raises ValueError if client is not found.
+    """
+    for config_entry_id in device.config_entries:
+        if client := hass.data[DOMAIN].get(config_entry_id):
+            return client
+
+    raise ValueError(
+        f"Device {device.id} is not from an existing {DOMAIN} config entry"
+    )
+
+
+@callback
+def async_get_device_id_from_entity_id(hass: HomeAssistant, entity_id: str) -> str:
+    """Get a device ID from an entity ID.
+
+    Raises ValueError is entity or device ID is invalid.
+    """
+    ent_reg = er.async_get(hass)
+    entity_entry = ent_reg.async_get(entity_id)
+
+    if (
+        entity_entry is None
+        or entity_entry.device_id is None
+        or entity_entry.platform != DOMAIN
+    ):
+        raise ValueError(f"Entity {entity_id} is not a valid {DOMAIN} entity.")
+
+    return entity_entry.device_id
```
Do we really need these helpers which are only used once?

splinter98 replied:
Not necessarily, but I was following the same pattern that is found in the [samsungtv](https://github.com/home-assistant/core/blob/b311fe2a0f6ff6de263cf5fb0fbef623d3304b0c/homeassistant/components/samsungtv/helpers.py#L16), and [webostv](https://github.com/home-assistant/core/blob/b311fe2a0f6ff6de263cf5fb0fbef623d3304b0c/homeassistant/components/webostv/helpers.py#L16) components which perform near-identical tasks.

So I didn't want to break existing convention, but it might be worth refactoring all of these helpers so that the core logic can be shared.

splinter98 replied:
I've now refactored the trigger helpers that are only used once

emontnemery commented on homeassistant/components/lg_netcast/manifest.json:
```diff
@@ -1,9 +1,12 @@
   "iot_class": "local_polling",
   "loggers": ["pylgnetcast"],
-  "requirements": ["pylgnetcast==0.3.7"]
+  "requirements": ["pylgnetcast==0.3.9"]
```
Does the dependency bump need to go hand in hand with adding the config flow? If not, please move the dependency bump to a separate PR.

splinter98 replied:
Yes, the dependency bump is here because we've added the `query_device_info` function that is used in the config_flow to get the UUID, model name and friendly name and of the TV.

emontnemery commented on homeassistant/components/lg_netcast/media_player.py:
```diff
@@ -54,23 +60,45 @@
+    )
 
-    add_entities([LgTVDevice(client, name, on_action_script)], True)
+    if (
+        result.get("type") == FlowResultType.ABORT
+        and result.get("reason") == "cannot_connect"
+    ):
+        raise PlatformNotReady(f"Connection error while connecting to {host}")
```
I think we should also raise an issue here, telling the user the import is pending because we can not connect to the TV and will be retried, the issue should be removed when import suceeds.

splinter98 replied:
This is handled in the config_flow.py

emontnemery replied:
OK, in that case I don't think we need to check the return value of the import flow here.

splinter98 replied:
Whilst the issue is now raised in the config_flow.py, the config_flow.py doesn't handle the waiting for the TV to be turned on and connected to the network to finish configuring the device. (This is because as part of the migration we look-up the UUID of the TV that's otherwise not present in the YAML config). Without the UUID the migration cannot proceed.

I'm using the PlatformNotReady Exception to trigger the  
Removing this check will require the user to restart Home Assistant with the TV online for the migration to occur. 

All other Aborts indicate an invalid or already migrated YAML configuration.

If this behaviour is undesirable or needs to be implemented in another way - let me know and I will adjust appropiately.

emontnemery replied:
Sure, that explanation makes perfect sense 👍

emontnemery commented on homeassistant/components/lg_netcast/media_player.py:
```diff
@@ -79,19 +107,39 @@ class LgTVDevice(MediaPlayerEntity):
+    _attr_name = None
 
-    def __init__(self, client, name, on_action_script):
+    def __init__(self, client, name, model, *, unique_id=None):
```
There's no reason the unique_id should be optional?
```suggestion
    def __init__(self, client, name, model, unique_id):
```

splinter98 replied:
Agreed - this is left from the original implementation that didn't migrate the YAML (and discover the unique_id)

splinter98 replied:
Done

emontnemery commented on homeassistant/components/lg_netcast/media_player.py:
```diff
@@ -79,19 +107,39 @@ class LgTVDevice(MediaPlayerEntity):
         self._program_name = ""
         self._sources = {}
         self._source_names = []
+        if unique_id is not None:
+            self._attr_unique_id = unique_id
+            self._attr_device_info = DeviceInfo(
+                identifiers={(DOMAIN, unique_id)},
+                manufacturer=ATTR_MANUFACTURER,
+                name=name,
+                model=model,
+            )
```
```suggestion
        self._attr_unique_id = unique_id
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, unique_id)},
            manufacturer=ATTR_MANUFACTURER,
            name=name,
            model=model,
        )
```

emontnemery commented on homeassistant/components/lg_netcast/media_player.py:
```diff
@@ -79,19 +107,39 @@ class LgTVDevice(MediaPlayerEntity):
+            )
+
+    async def async_added_to_hass(self) -> None:
+        """Connect and subscribe to dispatacher signals and state updates."""
```
```suggestion
        """Connect and subscribe to dispatcher signals and state updates."""
```

emontnemery commented on homeassistant/components/lg_netcast/media_player.py:
```diff
@@ -79,19 +107,39 @@ class LgTVDevice(MediaPlayerEntity):
+        """Connect and subscribe to dispatacher signals and state updates."""
+        await super().async_added_to_hass()
+
+        if (entry := self.registry_entry) and entry.device_id:
```
we know this entity has a unique id, hence there's no need to check this

splinter98 replied:
Sure, I think this is mainly here to keep the type checker happy (and is the same structure as the samgsung and webostv and webostv components) but I can instead turn this into a `TYPING_CHECKING` assert check instead.

splinter98 replied:
Done

emontnemery commented on homeassistant/components/lg_netcast/media_player.py:
```diff
@@ -211,8 +254,10 @@ def turn_off(self) -> None:
 
     def turn_on(self) -> None:
         """Turn on the media player."""
-        if self._on_action_script:
-            self._on_action_script.run(context=self._context)
+        asyncio.run_coroutine_threadsafe(
+            self._turn_on.async_run(self.hass, self._context),
+            self.hass.loop,
+        ).result()
```
If the turn on action is now a coroutine, this can be simplified to:
```
    async def async_turn_on(self) -> None:
        """Turn on the media player."""
        await self._turn_on.async_run(self.hass, self._context)
```

splinter98 committed:
Remove config_flow only consts and minor fixups

splinter98 committed:
Simplify media_player logic & raise new migration issue

splinter98 committed:
Add async_unload_entry

splinter98 committed:
Create issues when import config flow fails, and raise only a single yaml deprecation issue type

splinter98 commented on homeassistant/components/lg_netcast/__init__.py:
Done

splinter98 commented on homeassistant/components/lg_netcast/config_flow.py:
```diff
@@ -0,0 +1,213 @@
+    async def async_discover_client(self):
+        """Handle Discovery step."""
+        self.create_client()
+        assert self.client is not None
```
Done

splinter98 committed:
Remove single use trigger helpers

Drafteed commented:
@splinter98 Please make sure not to force push or amend commits after review has started to make it easier for readers to track changes.

splinter98 commented:
> @splinter98 Please make sure not to force push or amend commits after review has started to make it easier for readers to track changes.

My apologies, to help those who are tracking the changes the new changes start from [b7f06a0](https://github.com/home-assistant/core/pull/104913/commits/b7f06a0f8da5bafcc19b3fcd158674f191683ecf) - and just the diff of just new changes can be seen at [b7f06a0..27d8ce2](https://github.com/home-assistant/core/pull/104913/files/b7f06a0f8da5bafcc19b3fcd158674f191683ecf..27d8ce283f459bcff51a512bb5a80d4aa34dc39a)

emontnemery commented on homeassistant/components/lg_netcast/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+                self.hass,
+                HOMEASSISTANT_DOMAIN,
+                f"deprecated_yaml_{DOMAIN}",
+                breaks_in_ha_version="2024.6.0",
```
This needs to be bumped now
```suggestion
                breaks_in_ha_version="2024.11.0",
```

emontnemery commented on homeassistant/components/lg_netcast/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+                    self.hass,
+                    DOMAIN,
+                    "deprecated_yaml_import_issue_{err.reason}",
+                    breaks_in_ha_version="2024.6.0",
```
```suggestion
                    breaks_in_ha_version="2024.11.0",
```

emontnemery reviewed (changes requested):
Looks good!

Just a minor comment to address, then this is good to go 👍

splinter98 committed:
Bump issue deprecation breakage version

emontnemery reviewed (approved):
Thanks, @splinter98 👍

emontnemery committed:
Merge branch 'dev' into lgnetcast-configflow

emontnemery committed:
Lint
